src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs

```diff
@@ -871,9 +874,35 @@ End Property
 
                 await WaitWaiterAsync(workspace.ExportProvider);
 
-                // mutate solution
-                workspace.OnSolutionAdded(solution);
+                // block solution cralwer from processing.
+                var globalOperation = workspace.Services.GetService<IGlobalOperationNotificationService>();
+                using (var operation = globalOperation.Start("Block SolutionCrawler"))
+                {
+                    // mutate solution
+                    workspace.OnSolutionAdded(solution);
+
+                    // wait for workspace events to be all processed
+                    var workspaceWaiter = GetListenerProvider(workspace.ExportProvider).GetWaiter(FeatureAttribute.Workspace);
+                    await workspaceWaiter.CreateWaitTask();
+
+                    // now wait for semantic processor to finish
+                    var crawlerListener = (AsynchronousOperationListener)GetListenerProvider(workspace.ExportProvider).GetListener(FeatureAttribute.SolutionCrawler);
+
+                    // first, wait for first work to be queued.
+                    // asyncToken named "EnqueueWorkItemAsync" will be added after semantic propagation event queue is processed.
+                    //
+                    // since asyncToken doesn't distinguish whether (1) certain event is happened but all processed or (2) it never happened yet,
+                    // to check (1), we must wait for first item, and then wait for all items to be processed.
+                    await crawlerListener.WaitUntilConditionIsMetAsync(pendingTokens => pendingTokens.Any(token => token.Name == "EnqueueWorkItemAsync"));
```

Comment:
I'm unfamiliar with this: once the condition goes to true, it won't go to false until we release the global operation? i.e. there isn't a race where where it can start and stop right away?

Reply:
it will not go away until global operation is released. basically global operation is blocking the task from being processed.